[ONNX] Add setType from user into InferredType and Reliable in ConstantValueMap

titaiwangms opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #88622

`setType` API is not respected in current exporter because the graph-level shape type inference simply overrides every NOT ONNX Op shape we had from node-level shape type inference. To address this issue, this PR (1) makes custom Op with `setType` **reliable** in ConstantValueMap to secure its shape/type information in pass:  _C._jit_pass_onnx. (2) If an invalid Op with shape/type in pass: _C._jit_pass_onnx_graph_shape_type_inference(graph-level), we recognize it as reliable.

1. In #62856, The refactor in onnx.cpp made regression on custom Op, as that was the step we should update custom Op shape/type information into ConstantValueMap for remaining Ops.

2. Add another condition besides IsValidONNXNode for custom Op setType in shape_type_inference.cpp. If all the node output has shape (not all dynamic), we say it's custom set type.

3. ~However, this PR won't solve the [issue](https://github.com/pytorch/pytorch/issues/87738#issuecomment-1292831219) that in the node-level shape type inference, exporter invokes the warning in terms of the unknow custom Op, since we process its symbolic_fn after this warning, but it would have shape/type if setType is used correctly. And that will be left for another issue to solve. #84661~ Add `no_type_warning` in UpdateReliable() and it only warns if non ONNX node with no given type appears.

Fixes #81693
Fixes #87738 

NOTE: not confident of this not breaking anything. Please share your thoughts if there is a robust test on your mind.

titaiwangms committed:
[ONNX] Add setType from user into InferredType and Reliable in ConstantValueMap

[ghstack-poisoned]

AllenTiTaiWang commented on torch/csrc/jit/passes/onnx/shape_type_inference.cpp:
```diff
@@ -2041,6 +2040,17 @@ void ONNXShapeTypeInference(
+                 ->cast<TensorType>()
+                 ->sizes()
+                 .concrete_sizes()
+                 .has_value()) {
```
@BowenBao Is there a more elegant way to approach what I wanted to do here. Basically, a non-ONNX op which has shape/type information should be get in this block.

titaiwangms committed:
Update on "[ONNX] Add setType from user into InferredType and Reliable in ConstantValueMap"


`setType` API is not respected in current exporter because the graph-level shape type inference simply overrides every NOT ONNX Op shape we had from node-level shape type inference. To address this issue, this PR (1) makes custom Op with `setType` **reliable** in ConstantValueMap to secure its shape/type information in pass:  _C._jit_pass_onnx. (2) If an invalid Op with shape/type in pass: _C._jit_pass_onnx_graph_shape_type_inference(graph-level), we recognize it as reliable.

1. In #62856, The refactor in onnx.cpp made regression on custom Op, as that was the step we should update custom Op shape/type information into ConstantValueMap for remaining Ops.
2. Add another condition besides IsValidONNXNode for custom Op setType in shape_type_inference.cpp

test cases: #88623 
Fixes #81693
Fixed #87738 

NOTE: not confident of this not breaking anything. Please share your thoughts if there is a robust test on your mind.

[ghstack-poisoned]

titaiwangms committed:
Update on "[ONNX] Add setType from user into InferredType and Reliable in ConstantValueMap"


`setType` API is not respected in current exporter because the graph-level shape type inference simply overrides every NOT ONNX Op shape we had from node-level shape type inference. To address this issue, this PR (1) makes custom Op with `setType` **reliable** in ConstantValueMap to secure its shape/type information in pass:  _C._jit_pass_onnx. (2) If an invalid Op with shape/type in pass: _C._jit_pass_onnx_graph_shape_type_inference(graph-level), we recognize it as reliable.

1. In #62856, The refactor in onnx.cpp made regression on custom Op, as that was the step we should update custom Op shape/type information into ConstantValueMap for remaining Ops.
2. Add another condition besides IsValidONNXNode for custom Op setType in shape_type_inference.cpp

test cases: #88623 
Fixes #81693
Fixed #87738 

NOTE: not confident of this not breaking anything. Please share your thoughts if there is a robust test on your mind.

[ghstack-poisoned]

AllenTiTaiWang commented on torch/csrc/jit/passes/onnx/shape_type_inference.cpp:
```diff
@@ -1949,6 +1948,19 @@ void ONNXShapeTypeInference(
+  size_t all_has_value = 0;
+  for (auto node_output : n->outputs()) {
+    if (auto output_type = node_output->type()->cast<TensorType>()) {
+      if (output_type->sizes().concrete_sizes().has_value()) {
```
@BowenBao Is there a more elegant way to approach what I wanted to do here? Basically, a non-ONNX op which has shape/type information should be get in this block.

BowenBao replied:
I guess the `non-ONNX` part should be checked via looking at domain?
`concrete_sizes` might be too strict. IIRC it doesn't count if you have symbolic shape. I think you only need to differentiate from the default "empty" tensor type? I wonder if there are direct apis, but if not perhaps you can check by if rank is known.

It is important that this happens before merging shape from tracer, and after calling `setType`.
If the code is like `g.op(...).setType(...)`. I'm afraid both above conditions are not satisifed...

AllenTiTaiWang replied:
Like this?
```suggestion
      if (output_type->dim()) {
```

AllenTiTaiWang replied:
https://github.com/pytorch/pytorch/pull/88622#discussion_r1025891348

AllenTiTaiWang commented on torch/csrc/jit/passes/onnx.cpp:
```diff
@@ -326,10 +325,17 @@ void NodeToONNX(
+          // the user setType can be only accessed in this step, and it
+          // should be reliable.
           MergeInferredTypeAndSetMap(
-              outputs[i], old->type(), outputs[i]->type(), false);
```
Why is this function (`MergeInferredTypeAndSetMap`) required?

AllenTiTaiWang replied:
The reason #62856 brought `MergeInferredTypeAndSetMap` here is that we can have tracer shape as early as right after symbolic_fn processing. That is, if we didn't have `MergeInferredTypeAndSetMap`, the graph here would contain nodes without any type like the following:
```
%1 : Tensor = com.microsoft::Inverse(%0), scope: __main__.CustomInverse:: 
```

AllenTiTaiWang commented on torch/csrc/jit/passes/onnx.cpp:
```diff
@@ -326,10 +325,17 @@ void NodeToONNX(
           MergeInferredTypeAndSetMap(
-              outputs[i], old->type(), outputs[i]->type(), false);
+              outputs[i], old->type(), outputs[i]->type());
+          UpdateReliable(
```
Add paramter to control warning.

AllenTiTaiWang replied:
Done. Now the spammy warnings should be gone, if user set the type right.

titaiwangms committed:
Update on "[ONNX] Add setType from user into InferredType and Reliable in ConstantValueMap"


`setType` API is not respected in current exporter because the graph-level shape type inference simply overrides every NOT ONNX Op shape we had from node-level shape type inference. To address this issue, this PR (1) makes custom Op with `setType` **reliable** in ConstantValueMap to secure its shape/type information in pass:  _C._jit_pass_onnx. (2) If an invalid Op with shape/type in pass: _C._jit_pass_onnx_graph_shape_type_inference(graph-level), we recognize it as reliable.

1. In #62856, The refactor in onnx.cpp made regression on custom Op, as that was the step we should update custom Op shape/type information into ConstantValueMap for remaining Ops.
2. Add another condition besides IsValidONNXNode for custom Op setType in shape_type_inference.cpp
3. However, this PR won't solve the [issue](https://github.com/pytorch/pytorch/issues/87738#issuecomment-1292831219) that in the node-level shape type inference, exporter invokes the warning in terms of the unknow custom Op, since we process its symbolic_fn after this warning, but it would have shape/type if setType is used correctly. And that will be left for another issue to solve. #84661

test cases: #88623 
Fixes #81693
Fixes #87738 

NOTE: not confident of this not breaking anything. Please share your thoughts if there is a robust test on your mind.

[ghstack-poisoned]

titaiwangms committed:
Update on "[ONNX] Add setType from user into InferredType and Reliable in ConstantValueMap"


`setType` API is not respected in current exporter because the graph-level shape type inference simply overrides every NOT ONNX Op shape we had from node-level shape type inference. To address this issue, this PR (1) makes custom Op with `setType` **reliable** in ConstantValueMap to secure its shape/type information in pass:  _C._jit_pass_onnx. (2) If an invalid Op with shape/type in pass: _C._jit_pass_onnx_graph_shape_type_inference(graph-level), we recognize it as reliable.

1. In #62856, The refactor in onnx.cpp made regression on custom Op, as that was the step we should update custom Op shape/type information into ConstantValueMap for remaining Ops.
2. Add another condition besides IsValidONNXNode for custom Op setType in shape_type_inference.cpp
3. ~However, this PR won't solve the [issue](https://github.com/pytorch/pytorch/issues/87738#issuecomment-1292831219) that in the node-level shape type inference, exporter invokes the warning in terms of the unknow custom Op, since we process its symbolic_fn after this warning, but it would have shape/type if setType is used correctly. And that will be left for another issue to solve. #84661~ Add `no_type_warning` in UpdateReliable() and it only warns if non ONNX node with no given type appears.

test cases: #88623 
Fixes #81693
Fixes #87738 

NOTE: not confident of this not breaking anything. Please share your thoughts if there is a robust test on your mind.

[ghstack-poisoned]

titaiwangms committed:
Update on "[ONNX] Add setType from user into InferredType and Reliable in ConstantValueMap"


`setType` API is not respected in current exporter because the graph-level shape type inference simply overrides every NOT ONNX Op shape we had from node-level shape type inference. To address this issue, this PR (1) makes custom Op with `setType` **reliable** in ConstantValueMap to secure its shape/type information in pass:  _C._jit_pass_onnx. (2) If an invalid Op with shape/type in pass: _C._jit_pass_onnx_graph_shape_type_inference(graph-level), we recognize it as reliable.

1. In #62856, The refactor in onnx.cpp made regression on custom Op, as that was the step we should update custom Op shape/type information into ConstantValueMap for remaining Ops.
2. Add another condition besides IsValidONNXNode for custom Op setType in shape_type_inference.cpp
3. ~However, this PR won't solve the [issue](https://github.com/pytorch/pytorch/issues/87738#issuecomment-1292831219) that in the node-level shape type inference, exporter invokes the warning in terms of the unknow custom Op, since we process its symbolic_fn after this warning, but it would have shape/type if setType is used correctly. And that will be left for another issue to solve. #84661~ Add `no_type_warning` in UpdateReliable() and it only warns if non ONNX node with no given type appears.

test cases: #88623 
Fixes #81693
Fixes #87738 

NOTE: not confident of this not breaking anything. Please share your thoughts if there is a robust test on your mind.

[ghstack-poisoned]

AllenTiTaiWang commented on torch/csrc/jit/passes/onnx/shape_type_inference.cpp:
```diff
@@ -1949,6 +1949,29 @@ void ONNXShapeTypeInference(
+        // is set by user. On the other hand, if all of them are *, we take this
+        // node does not have given type, since unreliable nodes have * shape
+        // anyway.
+        for (size_t i = 0; i < output_type->dim().value(); i++) {
```
Only check `output_type->dim().has_value()` is not sufficient in graph-level shape type inference, as we already merged tracer shape right after symbolic_fn(the `MergeInferredTypeAndSetMap`), and every node now has dim (even an unreliable node has something like Float(*, *, *, *)). So I set another rule here, if the node has at least one static size (symbolic_sizes()[i].value() < 0 when it's *), it should be seen as type given. Otherwise, it really doesn't matter in terms of the whole dynamic sizes it has.

titaiwangms committed:
Update on "[ONNX] Add setType from user into InferredType and Reliable in ConstantValueMap"


`setType` API is not respected in current exporter because the graph-level shape type inference simply overrides every NOT ONNX Op shape we had from node-level shape type inference. To address this issue, this PR (1) makes custom Op with `setType` **reliable** in ConstantValueMap to secure its shape/type information in pass:  _C._jit_pass_onnx. (2) If an invalid Op with shape/type in pass: _C._jit_pass_onnx_graph_shape_type_inference(graph-level), we recognize it as reliable.

1. In #62856, The refactor in onnx.cpp made regression on custom Op, as that was the step we should update custom Op shape/type information into ConstantValueMap for remaining Ops.
2. Add another condition besides IsValidONNXNode for custom Op setType in shape_type_inference.cpp
3. ~However, this PR won't solve the [issue](https://github.com/pytorch/pytorch/issues/87738#issuecomment-1292831219) that in the node-level shape type inference, exporter invokes the warning in terms of the unknow custom Op, since we process its symbolic_fn after this warning, but it would have shape/type if setType is used correctly. And that will be left for another issue to solve. #84661~ Add `no_type_warning` in UpdateReliable() and it only warns if non ONNX node with no given type appears.

test cases: #88623 
Fixes #81693
Fixes #87738 

NOTE: not confident of this not breaking anything. Please share your thoughts if there is a robust test on your mind.

[ghstack-poisoned]

titaiwangms committed:
Update on "[ONNX] Add setType from user into InferredType and Reliable in ConstantValueMap"


`setType` API is not respected in current exporter because the graph-level shape type inference simply overrides every NOT ONNX Op shape we had from node-level shape type inference. To address this issue, this PR (1) makes custom Op with `setType` **reliable** in ConstantValueMap to secure its shape/type information in pass:  _C._jit_pass_onnx. (2) If an invalid Op with shape/type in pass: _C._jit_pass_onnx_graph_shape_type_inference(graph-level), we recognize it as reliable.

1. In #62856, The refactor in onnx.cpp made regression on custom Op, as that was the step we should update custom Op shape/type information into ConstantValueMap for remaining Ops.
2. Add another condition besides IsValidONNXNode for custom Op setType in shape_type_inference.cpp
3. ~However, this PR won't solve the [issue](https://github.com/pytorch/pytorch/issues/87738#issuecomment-1292831219) that in the node-level shape type inference, exporter invokes the warning in terms of the unknow custom Op, since we process its symbolic_fn after this warning, but it would have shape/type if setType is used correctly. And that will be left for another issue to solve. #84661~ Add `no_type_warning` in UpdateReliable() and it only warns if non ONNX node with no given type appears.

test cases: #88623 
Fixes #81693
Fixes #87738 

NOTE: not confident of this not breaking anything. Please share your thoughts if there is a robust test on your mind.

[ghstack-poisoned]

titaiwangms committed:
Update on "[ONNX] Add setType from user into InferredType and Reliable in ConstantValueMap"


`setType` API is not respected in current exporter because the graph-level shape type inference simply overrides every NOT ONNX Op shape we had from node-level shape type inference. To address this issue, this PR (1) makes custom Op with `setType` **reliable** in ConstantValueMap to secure its shape/type information in pass:  _C._jit_pass_onnx. (2) If an invalid Op with shape/type in pass: _C._jit_pass_onnx_graph_shape_type_inference(graph-level), we recognize it as reliable.

1. In #62856, The refactor in onnx.cpp made regression on custom Op, as that was the step we should update custom Op shape/type information into ConstantValueMap for remaining Ops.
2. Add another condition besides IsValidONNXNode for custom Op setType in shape_type_inference.cpp
3. ~However, this PR won't solve the [issue](https://github.com/pytorch/pytorch/issues/87738#issuecomment-1292831219) that in the node-level shape type inference, exporter invokes the warning in terms of the unknow custom Op, since we process its symbolic_fn after this warning, but it would have shape/type if setType is used correctly. And that will be left for another issue to solve. #84661~ Add `no_type_warning` in UpdateReliable() and it only warns if non ONNX node with no given type appears.

test cases: #88623 
Fixes #81693
Fixes #87738 

NOTE: not confident of this not breaking anything. Please share your thoughts if there is a robust test on your mind.

[ghstack-poisoned]

BowenBao commented on test/onnx/test_pytorch_onnx_shape_inference.py:
```diff
@@ -284,5 +286,171 @@ def test_reduce_prod_without_axes(self):
         self.run_test(g, reduce_prod.node(), expect_tensor("Long", shape=(1,)))
 
 
+class TestONNXCustomOpShapeInference(common_utils.TestCase):
```
```suggestion
class TestONNXCustomOpShapeInference(pytorch_test_common.ExportTestCase):
```

AllenTiTaiWang replied:
Done

BowenBao commented on test/onnx/test_pytorch_onnx_shape_inference.py:
```diff
@@ -284,5 +286,171 @@ def test_reduce_prod_without_axes(self):
 
+class TestONNXCustomOpShapeInference(common_utils.TestCase):
+
+    opset_version = _constants.ONNX_MAX_OPSET
```
Consider putting this in setUp
```python
def setUp(self):
    super().setUp()
    self.opset_version = _constants.ONNX_MAX_OPSET
```

AllenTiTaiWang replied:
Done

BowenBao commented on test/onnx/test_pytorch_onnx_shape_inference.py:
```diff
@@ -284,5 +286,171 @@ def test_reduce_prod_without_axes(self):
+        model_proto = onnx.load(io.BytesIO(f.getvalue()))
+        # To validate the shape of inverse Op, we need to find inverse output name,
+        # and then use it to identify its value_info for the shape.
+        output_name = ""
```
Should we assert on `output_name` not empty str after the loop? It's a bit paranoid but this test will silently pass if `Inverse` is not found in exported graph, as the rest of asserts are mostly skipped from not entering `if value_info.name == output_name:`

AllenTiTaiWang replied:
Added assert output_name

BowenBao commented on torch/csrc/jit/passes/onnx/shape_type_inference.cpp:
```diff
@@ -1949,6 +1949,29 @@ void ONNXShapeTypeInference(
+
+  // Go through every output to check if they all have shape
+  // If they all do, this should be reliable even if the Op is not from ONNX.
+  size_t all_output_has_value = 0;
```
nit: variable name implies bool rather than count.

AllenTiTaiWang replied:
Done

BowenBao commented on torch/csrc/jit/passes/onnx/shape_type_inference.cpp:
```diff
@@ -1949,6 +1949,29 @@ void ONNXShapeTypeInference(
+        // node does not have given type, since unreliable nodes have * shape
+        // anyway.
+        for (size_t i = 0; i < output_type->dim().value(); i++) {
+          if (output_type->symbolic_sizes()[i].value() > 0) {
```
nit: wonder if can simplify the loop to just check if has `symbolic_sizes`, and if so loop over it.

AllenTiTaiWang replied:
Found and used is_static() and symbolic_sizes().sizes(), and they come in handy.

BowenBao commented on torch/csrc/jit/passes/onnx/shape_type_inference.cpp:
```diff
@@ -2041,6 +2064,12 @@ void ONNXShapeTypeInference(
       GRAPH_DEBUG(
           "ONNX graph after shape inference: ", prettyPrint(*model_proto));
     }
+  } else if (custom_setType) {
```
nit: consider make a function/lambda to get custom_setType, to "lazy execute" it only if needed (when getting to this branch)

AllenTiTaiWang replied:
Done

BowenBao reviewed (approved):
LGTM, thanks for deep diving to this issue! I left some inline suggestions.
The reliable system is not very reliable :). We may need to revisit this.

titaiwangms committed:
Update on "[ONNX] Add setType from user into InferredType and Reliable in ConstantValueMap"


`setType` API is not respected in current exporter because the graph-level shape type inference simply overrides every NOT ONNX Op shape we had from node-level shape type inference. To address this issue, this PR (1) makes custom Op with `setType` **reliable** in ConstantValueMap to secure its shape/type information in pass:  _C._jit_pass_onnx. (2) If an invalid Op with shape/type in pass: _C._jit_pass_onnx_graph_shape_type_inference(graph-level), we recognize it as reliable.

1. In #62856, The refactor in onnx.cpp made regression on custom Op, as that was the step we should update custom Op shape/type information into ConstantValueMap for remaining Ops.

2. Add another condition besides IsValidONNXNode for custom Op setType in shape_type_inference.cpp. If all the node output has shape (not all dynamic), we say it's custom set type.

3. ~However, this PR won't solve the [issue](https://github.com/pytorch/pytorch/issues/87738#issuecomment-1292831219) that in the node-level shape type inference, exporter invokes the warning in terms of the unknow custom Op, since we process its symbolic_fn after this warning, but it would have shape/type if setType is used correctly. And that will be left for another issue to solve. #84661~ Add `no_type_warning` in UpdateReliable() and it only warns if non ONNX node with no given type appears.

Fixes #81693
Fixes #87738 

NOTE: not confident of this not breaking anything. Please share your thoughts if there is a robust test on your mind.

[ghstack-poisoned]

titaiwangms committed:
Update on "[ONNX] Add setType from user into InferredType and Reliable in ConstantValueMap"


`setType` API is not respected in current exporter because the graph-level shape type inference simply overrides every NOT ONNX Op shape we had from node-level shape type inference. To address this issue, this PR (1) makes custom Op with `setType` **reliable** in ConstantValueMap to secure its shape/type information in pass:  _C._jit_pass_onnx. (2) If an invalid Op with shape/type in pass: _C._jit_pass_onnx_graph_shape_type_inference(graph-level), we recognize it as reliable.

1. In #62856, The refactor in onnx.cpp made regression on custom Op, as that was the step we should update custom Op shape/type information into ConstantValueMap for remaining Ops.

2. Add another condition besides IsValidONNXNode for custom Op setType in shape_type_inference.cpp. If all the node output has shape (not all dynamic), we say it's custom set type.

3. ~However, this PR won't solve the [issue](https://github.com/pytorch/pytorch/issues/87738#issuecomment-1292831219) that in the node-level shape type inference, exporter invokes the warning in terms of the unknow custom Op, since we process its symbolic_fn after this warning, but it would have shape/type if setType is used correctly. And that will be left for another issue to solve. #84661~ Add `no_type_warning` in UpdateReliable() and it only warns if non ONNX node with no given type appears.

Fixes #81693
Fixes #87738 

NOTE: not confident of this not breaking anything. Please share your thoughts if there is a robust test on your mind.

[ghstack-poisoned]

AllenTiTaiWang commented:
@pytorchbot merge -g

AllenTiTaiWang commented:
> LGTM, thanks for deep diving to this issue! I left some inline suggestions. The reliable system is not very reliable :). We may need to revisit this.

When we can leverage on PyTorch graph shape type inference, we won't need this anymore..

AllenTiTaiWang commented:
@pytorchbot merge -g
